File: applications/rtkfdk/rtkfdk.cxx
/*=========================================================================
 *
 *  Copyright RTK Consortium
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *         https://www.apache.org/licenses/LICENSE-2.0.txt
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 *=========================================================================*/

#include "rtkfdk_ggo.h"
#include "rtkGgoFunctions.h"
#include "rtkConfiguration.h"

#include "rtkThreeDCircularProjectionGeometryXMLFile.h"
#include "rtkDisplacedDetectorForOffsetFieldOfViewImageFilter.h"
#include "rtkParkerShortScanImageFilter.h"
#include "rtkFDKConeBeamReconstructionFilter.h"
#ifdef RTK_USE_CUDA
#  include "rtkCudaDisplacedDetectorImageFilter.h"
// TODO #  include "rtkCudaDisplacedDetectorForOffsetFieldOfViewImageFilter.h"
#  include "rtkCudaParkerShortScanImageFilter.h"
#  include "rtkCudaFDKConeBeamReconstructionFilter.h"
#endif
#include "rtkFDKWarpBackProjectionImageFilter.h"
#include "rtkCyclicDeformationImageFilter.h"
#include "rtkProgressCommands.h"

#include <itkStreamingImageFilter.h>
#include <itkImageRegionSplitterDirection.h>
#include <itkImageFileWriter.h>

int
main(int argc, char * argv[])
{
  GGO(rtkfdk, args_info);

  using OutputPixelType = float;
  constexpr unsigned int Dimension = 3;

  using CPUOutputImageType = itk::Image<OutputPixelType, Dimension>;
#ifdef RTK_USE_CUDA
  using OutputImageType = itk::CudaImage<OutputPixelType, Dimension>;
#else
  using OutputImageType = CPUOutputImageType;
#endif

  // Projections reader
  using ReaderType = rtk::ProjectionsReader<OutputImageType>;
  auto reader = ReaderType::New();
  rtk::SetProjectionsReaderFromGgo<ReaderType, args_info_rtkfdk>(reader, args_info);

  if (!args_info.lowmem_flag)
  {
    if (args_info.verbose_flag)
      std::cout << "Reading... " << std::endl;
    TRY_AND_EXIT_ON_ITK_EXCEPTION(reader->Update())
  }

  // Geometry
  if (args_info.verbose_flag)
    std::cout << "Reading geometry information from " << args_info.geometry_arg << "..." << std::endl;
  rtk::ThreeDCircularProjectionGeometry::Pointer geometry;
  TRY_AND_EXIT_ON_ITK_EXCEPTION(geometry = rtk::ReadGeometry(args_info.geometry_arg));

  // Check on hardware parameter
#ifndef RTK_USE_CUDA
  if (!strcmp(args_info.hardware_arg, "cuda"))
  {
    std::cerr << "The program has not been compiled with cuda option" << std::endl;
    return EXIT_FAILURE;
  }
#endif

  // Displaced detector weighting
#ifdef RTK_USE_CUDA
  using DDFType = rtk::CudaDisplacedDetectorImageFilter;
#else
  using DDFType = rtk::DisplacedDetectorForOffsetFieldOfViewImageFilter<OutputImageType>;
#endif
  rtk::DisplacedDetectorImageFilter<OutputImageType>::Pointer ddf;
  if (!strcmp(args_info.hardware_arg, "cuda"))
    ddf = DDFType::New();
  else
    ddf = rtk::DisplacedDetectorForOffsetFieldOfViewImageFilter<OutputImageType>::New();
  ddf->SetInput(reader->GetOutput());
  ddf->SetGeometry(geometry);
  ddf->SetDisable(args_info.nodisplaced_flag);

  // Short scan image filter
  using PSSFCPUType = rtk::ParkerShortScanImageFilter<OutputImageType>;
#ifdef RTK_USE_CUDA
  using PSSFType = rtk::CudaParkerShortScanImageFilter;
#else
  using PSSFType = rtk::ParkerShortScanImageFilter<OutputImageType>;
#endif
  PSSFCPUType::Pointer pssf;
  if (!strcmp(args_info.hardware_arg, "cuda"))
    pssf = PSSFType::New();
  else
    pssf = PSSFCPUType::New();
  pssf->SetInput(ddf->GetOutput());
  pssf->SetGeometry(geometry);
  pssf->InPlaceOff();
  pssf->SetAngularGapThreshold(args_info.short_arg * itk::Math::pi / 180.);

  // Create reconstructed image
  using ConstantImageSourceType = rtk::ConstantImageSource<OutputImageType>;
  auto constantImageSource = ConstantImageSourceType::New();
  rtk::SetConstantImageSourceFromGgo<ConstantImageSourceType, args_info_rtkfdk>(constantImageSource, args_info);

  // Motion-compensated objects for the compensation of a cyclic deformation.
  // Although these will only be used if the command line options for motion
  // compensation are set, we still create the object before hand to avoid auto
  // destruction.
  using DVFPixelType = itk::Vector<float, 3>;
  using DeformationType = rtk::CyclicDeformationImageFilter<itk::Image<DVFPixelType, 4>, itk::Image<DVFPixelType, 3>>;
  auto dvfReader = itk::ImageFileReader<DeformationType::InputImageType>::New();
  auto def = DeformationType::New();
  def->SetInput(dvfReader->GetOutput());
  auto bp = rtk::FDKWarpBackProjectionImageFilter<OutputImageType, OutputImageType, DeformationType>::New();
  bp->SetDeformation(def);
  bp->SetGeometry(geometry);

  // This macro sets options for fdk filter which I can not see how to do better
  // because TFFTPrecision is not the same, e.g. for CPU and CUDA (SR)
#define SET_FELDKAMP_OPTIONS(f)                                   \
  f->SetInput(0, constantImageSource->GetOutput());               \
  f->SetInput(1, pssf->GetOutput());                              \
  f->SetGeometry(geometry);                                       \
  f->GetRampFilter()->SetTruncationCorrection(args_info.pad_arg); \
  f->GetRampFilter()->SetHannCutFrequency(args_info.hann_arg);    \
  f->GetRampFilter()->SetHannCutFrequencyY(args_info.hannY_arg);  \
  f->SetProjectionSubsetSize(args_info.subsetsize_arg);           \
  if (args_info.verbose_flag)                                     \
  {                                                               \
    f->AddObserver(itk::AnyEvent(), progressCommand);             \
  }

  // FDK reconstruction filtering
  using FDKCPUType = rtk::FDKConeBeamReconstructionFilter<OutputImageType>;
  FDKCPUType::Pointer feldkamp;
#ifdef RTK_USE_CUDA
  using FDKCUDAType = rtk::CudaFDKConeBeamReconstructionFilter;
  FDKCUDAType::Pointer feldkampCUDA;
#endif

  itk::Image<OutputPixelType, Dimension> * pfeldkamp = nullptr;
  if (!strcmp(args_info.hardware_arg, "cpu"))
  {
    // Progress reporting
    using PercentageProgressCommandType = rtk::PercentageProgressCommand<FDKCPUType>;
    auto progressCommand = PercentageProgressCommandType::New();

    feldkamp = FDKCPUType::New();
    SET_FELDKAMP_OPTIONS(feldkamp);

    // Motion compensated CBCT settings
    if (args_info.signal_given && args_info.dvf_given)
    {
      dvfReader->SetFileName(args_info.dvf_arg);
      def->SetSignalFilename(args_info.signal_arg);
      feldkamp->SetBackProjectionFilter(bp.GetPointer());
    }
    pfeldkamp = feldkamp->GetOutput();
  }
#ifdef RTK_USE_CUDA
  else if (!strcmp(args_info.hardware_arg, "cuda"))
  {
    // Motion compensation not supported in cuda
    if (args_info.signal_given && args_info.dvf_given)
    {
      std::cerr << "Motion compensation is not supported in CUDA. Aborting" << std::endl;
      return EXIT_FAILURE;
    }

    // Progress reporting
    using PercentageProgressCommandType = rtk::PercentageProgressCommand<FDKCUDAType>;
    auto progressCommand = PercentageProgressCommandType::New();

    feldkampCUDA = FDKCUDAType::New();
    SET_FELDKAMP_OPTIONS(feldkampCUDA);
    pfeldkamp = feldkampCUDA->GetOutput();
  }
#endif

  // Streaming depending on streaming capability of writer
  auto streamerBP = itk::StreamingImageFilter<CPUOutputImageType, CPUOutputImageType>::New();
  streamerBP->SetInput(pfeldkamp);
  streamerBP->SetNumberOfStreamDivisions(args_info.divisions_arg);
  auto splitter = itk::ImageRegionSplitterDirection::New();
  splitter->SetDirection(2); // Prevent splitting along z axis. As a result, splitting will be performed along y axis
  streamerBP->SetRegionSplitter(splitter);

  // Write
  if (args_info.verbose_flag)
    std::cout << "Reconstructing and writing... " << std::endl;

  TRY_AND_EXIT_ON_ITK_EXCEPTION(itk::WriteImage(streamerBP->GetOutput(), args_info.output_arg))

  return EXIT_SUCCESS;
}

File: applications/rtkfdk/rtkfdk.py
#!/usr/bin/env python
import argparse
import sys
import math
import itk
from itk import RTK as rtk
import shlex


def build_parser():
    parser = argparse.ArgumentParser(
        description="Reconstructs a 3D volume from a sequence of projections [Feldkamp, David, Kress, 1984].",
        formatter_class=argparse.ArgumentDefaultsHelpFormatter,
    )

    # General options
    parser.add_argument(
        "--verbose", "-v", help="Verbose execution", action="store_true"
    )
    parser.add_argument(
        "--geometry", "-g", help="XML geometry file name", type=str, required=True
    )
    parser.add_argument(
        "--output", "-o", help="Output file name", type=str, required=True
    )
    parser.add_argument(
        "--hardware",
        help="Hardware used for computation",
        choices=["cpu", "cuda"],
        default="cpu",
    )
    parser.add_argument(
        "--lowmem",
        "-l",
        help="Load only one projection per thread in memory",
        action="store_true",
    )
    parser.add_argument(
        "--divisions",
        "-d",
        help="Streaming option: number of stream divisions of the CT",
        type=int,
        default=1,
    )
    parser.add_argument(
        "--subsetsize",
        help="Streaming option: number of projections processed at a time",
        type=int,
        default=16,
    )
    parser.add_argument(
        "--nodisplaced",
        help="Disable the displaced detector filter",
        action="store_true",
    )
    parser.add_argument(
        "--short",
        help="Minimum angular gap to detect a short scan (converted to radians)",
        type=float,
        default=20.0,
    )

    # Ramp filter options
    ramp_group = parser.add_argument_group("Ramp filter")
    ramp_group.add_argument(
        "--pad",
        help="Data padding parameter to correct for truncation",
        type=float,
        default=0.0,
    )
    ramp_group.add_argument(
        "--hann",
        help="Cut frequency for Hann window in ]0,1] (0.0 disables it)",
        type=float,
        default=0.0,
    )
    ramp_group.add_argument(
        "--hannY",
        help="Cut frequency for Hann window in ]0,1] (0.0 disables it)",
        type=float,
        default=0.0,
    )

    # Motion compensation options
    motion_group = parser.add_argument_group(
        "Motion compensation described in [Rit et al, TMI, 2009] and [Rit et al, Med Phys, 2009]"
    )
    motion_group.add_argument("--signal", help="Signal file name", type=str)
    motion_group.add_argument("--dvf", help="Input 4D DVF", type=str)

    # RTK specific groups
    rtk.add_rtkinputprojections_group(parser)
    rtk.add_rtk3Doutputimage_group(parser)

    # Parse the command line arguments
    return parser


def process(args: argparse.Namespace):
    # Define pixel type and dimension
    OutputPixelType = itk.F
    Dimension = 3
    OutputImageType = itk.Image[OutputPixelType, Dimension]

    # Projections reader
    reader = rtk.ProjectionsReader[OutputImageType].New()
    rtk.SetProjectionsReaderFromArgParse(reader, args)

    if not args.lowmem:
        if args.verbose:
            print("Reading projections...")
        reader.Update()

    # Geometry
    if args.verbose:
        print(f"Reading geometry information from {args.geometry}...")
    geometry = rtk.read_geometry(args.geometry)

    # Check on hardware parameter
    if not hasattr(itk, "CudaImage") and args.hardware == "cuda":
        print("The program has not been compiled with CUDA option.")
        sys.exit(1)

    # CUDA classes are non-templated, so they do not require a type specification.
    if args.hardware == "cuda":
        ddf = rtk.CudaDisplacedDetectorImageFilter.New()
        ddf.SetInput(itk.cuda_image_from_image(reader.GetOutput()))
        pssf = rtk.CudaParkerShortScanImageFilter.New()
    else:
        ddf = rtk.DisplacedDetectorForOffsetFieldOfViewImageFilter[
            OutputImageType
        ].New()
        ddf.SetInput(reader.GetOutput())
        pssf = rtk.ParkerShortScanImageFilter[OutputImageType].New()

    # Displaced detector weighting
    ddf.SetGeometry(geometry)
    ddf.SetDisable(args.nodisplaced)

    # Short scan image filter
    pssf.SetInput(ddf.GetOutput())
    pssf.SetGeometry(geometry)
    pssf.InPlaceOff()
    pssf.SetAngularGapThreshold(args.short * math.pi / 180.0)

    # Create reconstructed image
    constantImageSource = rtk.ConstantImageSource[OutputImageType].New()
    rtk.SetConstantImageSourceFromArgParse(constantImageSource, args)

    # Motion-compensated objects for the compensation of a cyclic deformation.
    # Although these will only be used if the command line options for motion
    # compensation are set, we still create the object beforehand to avoid auto
    # destruction.
    DVFPixelType = itk.Vector[itk.F, 3]
    DVFImageType = itk.Image[DVFPixelType, 3]
    DVFImageSequenceType = itk.Image[DVFPixelType, 4]

    # Create the deformation filter and reader for DVF
    deformation = rtk.CyclicDeformationImageFilter[
        DVFImageSequenceType, DVFImageType
    ].New()
    dvfReader = itk.ImageFileReader[DVFImageSequenceType].New()
    deformation.SetInput(dvfReader.GetOutput())

    # Set up the back projection filter for motion compensation
    bp = rtk.FDKWarpBackProjectionImageFilter[
        OutputImageType, OutputImageType, type(deformation)
    ].New()
    bp.SetDeformation(deformation)
    bp.SetGeometry(geometry)

    # FDK reconstruction filtering
    if args.hardware == "cuda":
        feldkamp = rtk.CudaFDKConeBeamReconstructionFilter.New()
        feldkamp.SetInput(0, itk.cuda_image_from_image(constantImageSource.GetOutput()))
    else:
        feldkamp = rtk.FDKConeBeamReconstructionFilter[OutputImageType].New()
        feldkamp.SetInput(0, constantImageSource.GetOutput())

    # Set inputs and options for the FDK filter
    feldkamp.SetInput(1, pssf.GetOutput())
    feldkamp.SetGeometry(geometry)
    feldkamp.GetRampFilter().SetTruncationCorrection(args.pad)
    feldkamp.GetRampFilter().SetHannCutFrequency(args.hann)
    feldkamp.GetRampFilter().SetHannCutFrequencyY(args.hannY)
    feldkamp.SetProjectionSubsetSize(args.subsetsize)

    # Progress reporting
    if args.verbose:
        progressCommand = rtk.PercentageProgressCommand(feldkamp)
        feldkamp.AddObserver(itk.ProgressEvent(), progressCommand.callback)
        feldkamp.AddObserver(itk.EndEvent(), progressCommand.End)

    # Motion compensated CBCT settings
    if args.signal and args.dvf:
        if args.hardware == "cuda":
            print("Motion compensation is not supported in CUDA. Aborting")
            sys.exit(1)  # Exit if CUDA is selected with motion compensation
        dvfReader.SetFileName(args.dvf)
        deformation.SetSignalFilename(args.signal)
        feldkamp.SetBackProjectionFilter(bp)

    # Streaming depending on streaming capability of writer
    streamerBP = itk.StreamingImageFilter[OutputImageType, OutputImageType].New()
    streamerBP.SetInput(feldkamp.GetOutput())
    streamerBP.SetNumberOfStreamDivisions(args.divisions)

    # Create a splitter to control how the image region is divided during streaming
    splitter = itk.ImageRegionSplitterDirection.New()
    splitter.SetDirection(2)
    streamerBP.SetRegionSplitter(splitter)

    if args.verbose:
        print(f"Reconstructing and writing...")

    # Write
    itk.imwrite(streamerBP.GetOutput(), args.output)


def main(argv=None):
    parser = build_parser()
    args = parser.parse_args(argv)
    process(args)


def rtkfdk(*args, **kwargs):
    """
    Unified RTKFDK entry point.

    Usage:
      • Shell-style: rtkfdk("-p . -r proj.mha -g geom.xml -o out.mha")
      • Python API:  rtkfdk(path=".", regexp="proj.mha", geometry="geom.xml", output="out.mha")
    """
    # Shell-style style
    if len(args) == 1 and isinstance(args[0], str) and not kwargs:
        argv = shlex.split(args[0])
        return main(argv)

    # Python API style
    if len(args) == 0:
        parser = build_parser()
        parsed = rtk.parse_kwargs(parser, func_name="rtkfdk", **kwargs)
        return process(parsed)


rtk.patch_signature(rtkfdk, build_parser())

if __name__ == "__main__":
    main()
